src/subscription_manager/utils.py

```diff
@@ -62,6 +62,76 @@ def __repr__(self):
         return pprint.pformat(self.as_dict())
 
 
+def three_way_merge(local=None, base=None, remote=None, on_conflict="remote"):
+    """
+    Performs a three-way merge on the local and remote dictionaries with a given base.
+    :param local: The dictionary of the current local values
+    :param base: The dictionary with the values we've last seen
+    :param remote: The dictionary with "their" values
+    :param winner_on_conflict: Either "remote" or "local" or None. If "remote", the remote changes
+                               will win any conflict. If "local", the local changes will win any
+                               conflict. If anything else, an error will be thrown.
+
+    :return: The dictionary of values as merged between the three provided dictionaries.
+    """
+    # Ensure that we at least have a value for each keyword arg. Left each input param as a keyword
+    # arg as it improves readability
+
+    result = {}
+
+    local = local or {}
+    base = base or {}
+    remote = remote or {}
+
+    if on_conflict == "remote":
+        winner = remote
+    elif on_conflict == "local":
+        winner = local
+    else:
+        winner = {}
+
+    all_keys = sorted(list(set(local.keys()) | set(base.keys()) | set(remote.keys())))
+
+    # An object used to represent a value not being set at all.
+    no_val = object()
+
+    for key in all_keys:
+
+        in_local = key in local
+        in_remote = key in remote
+        in_base = key in base
+
+        local_value = local.get(key, no_val)
+        remote_value = remote.get(key, no_val)
+        base_value = base.get(key, no_val)
+
+        if not in_base:
+            if in_local and in_remote:
+                # Either conflicting addition or non-conflicting addition
+                result[key] = winner[key]
+            elif in_local and not in_remote:
+                # Added only in local
+                result[key] = local[key]
+            elif in_remote and not in_local:
+                # Added only in remote
+                result[key] = remote[key]
+        elif in_base:
+            if in_local != in_remote:
+                if key in winner:
+                    result[key] = winner[key]
+            elif in_local and local_value != base_value:
+                if in_remote and remote_value != local_value:
+                    result[key] = winner[key]
+                else:
+                    result[key] = local[key]
+            elif in_remote and remote_value != base_value:
+                result[key] = remote[key]
+            elif in_local and in_remote and local_value == remote_value and remote_value == base_value:
+                result[key] = base[key]
```

Comment:
This if-elif-else part of code is little bit wild. We can have only following cases, when key is included in base dictionary:

|  local (winner) | base | remote | result |
|-----------------|------|--------|--------|
|                 | B    |        |        |
| A               | A    |        | A      |
| A               | B    |        | A      |
|                 | B    | B      |        |
|                 | B    | C      |        |
| A               | A    | A      | A      |
| A               | B    | A      | A      |
| A               | B    | C      | A      |

See that cases in table are sorted from simple to more complicates. Thus, I would simplify code to something like this:

```python
if local == winner:
    if in_local:
        # it doesn't matter if local[key] equals to base[key] or not,
        # because local is winner in this case
        result[key] = local[key]
elif remote == winner:
    if in_remote:
        # it doesn't matter if remote[key] equals to base[key] or not,
        # because remote is winner in this case
        result[key] = remote[key]
# When winner could not be anything else then "remote/local", then following code needn't to be here
else:
    if local[key] != remote[key]:
        if key in winner:
            result[key] = winner[key]
        else:
            raise ValueError('Conflict and no winner specified')
    else:
        result[key] = local[key]  # same as remote[key]
```